{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Resolución de ambigüedades taxonID\n",
    "\n",
    "Se buscará mejorar la resolución de ambigüedades en el filtrado de GSMs buscando el txid más específico dentro de los campos de GPL_txid y GSM_ch1_characteristics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dependencias \n",
    "from Bio import Entrez\n",
    "import GEOparse\n",
    "import pandas as pd\n",
    "import re\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv(\"../230827FilterFunc/bigprueba/27082023Test.tsv\", sep=\"\\t\").rename(columns={\"taxonID\":\"TaxId\"})\n",
    "data.TaxId = data.TaxId.astype(\"str\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [],
   "source": [
    "rel = pd.read_csv(\"../230903MultiPrEstable/tests/taxonomy_relations_mini.tsv\", sep='\\t')\n",
    "rel.TaxId = rel.TaxId.astype(\"str\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = data.merge(rel, on=\"TaxId\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [],
   "source": [
    "no_match = data.iloc[[i for i,row in data.iterrows() if row.TaxId != row.GSM_taxid_ch1]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TaxId</th>\n",
       "      <th>GSM_taxid_ch1</th>\n",
       "      <th>GPL_taxid</th>\n",
       "      <th>GPL_organism</th>\n",
       "      <th>GSM_characteristics_ch1</th>\n",
       "      <th>GSM_organism_ch1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>224308</td>\n",
       "      <td>1718</td>\n",
       "      <td>1718;160488;224308;290633;511145</td>\n",
       "      <td>Corynebacterium glutamicum;Pseudomonas putida ...</td>\n",
       "      <td>genotype: ChrS-Ala245fs</td>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>224308</td>\n",
       "      <td>1718</td>\n",
       "      <td>1718;160488;224308;290633;511145</td>\n",
       "      <td>Corynebacterium glutamicum;Pseudomonas putida ...</td>\n",
       "      <td>genotype: ChrS-Ala245fs</td>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>224308</td>\n",
       "      <td>1718</td>\n",
       "      <td>1718;160488;224308;290633;511145</td>\n",
       "      <td>Corynebacterium glutamicum;Pseudomonas putida ...</td>\n",
       "      <td>genotype: ChrS-Ala245fs</td>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>224308</td>\n",
       "      <td>1423</td>\n",
       "      <td>224308</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>224308</td>\n",
       "      <td>1423</td>\n",
       "      <td>224308</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13646</th>\n",
       "      <td>1314</td>\n",
       "      <td>160490</td>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13647</th>\n",
       "      <td>1314</td>\n",
       "      <td>160490</td>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13648</th>\n",
       "      <td>1314</td>\n",
       "      <td>160490</td>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13649</th>\n",
       "      <td>1314</td>\n",
       "      <td>160490</td>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13650</th>\n",
       "      <td>1314</td>\n",
       "      <td>160490</td>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7756 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        TaxId GSM_taxid_ch1                         GPL_taxid  \\\n",
       "0      224308          1718  1718;160488;224308;290633;511145   \n",
       "1      224308          1718  1718;160488;224308;290633;511145   \n",
       "2      224308          1718  1718;160488;224308;290633;511145   \n",
       "3      224308          1423                            224308   \n",
       "4      224308          1423                            224308   \n",
       "...       ...           ...                               ...   \n",
       "13646    1314        160490                              1314   \n",
       "13647    1314        160490                              1314   \n",
       "13648    1314        160490                              1314   \n",
       "13649    1314        160490                              1314   \n",
       "13650    1314        160490                              1314   \n",
       "\n",
       "                                            GPL_organism  \\\n",
       "0      Corynebacterium glutamicum;Pseudomonas putida ...   \n",
       "1      Corynebacterium glutamicum;Pseudomonas putida ...   \n",
       "2      Corynebacterium glutamicum;Pseudomonas putida ...   \n",
       "3             Bacillus subtilis subsp. subtilis str. 168   \n",
       "4             Bacillus subtilis subsp. subtilis str. 168   \n",
       "...                                                  ...   \n",
       "13646                             Streptococcus pyogenes   \n",
       "13647                             Streptococcus pyogenes   \n",
       "13648                             Streptococcus pyogenes   \n",
       "13649                             Streptococcus pyogenes   \n",
       "13650                             Streptococcus pyogenes   \n",
       "\n",
       "                                 GSM_characteristics_ch1  \\\n",
       "0                                genotype: ChrS-Ala245fs   \n",
       "1                                genotype: ChrS-Ala245fs   \n",
       "2                                genotype: ChrS-Ala245fs   \n",
       "3      strain: NDmed;genotype: Wild-type, undomestica...   \n",
       "4      strain: NDmed;genotype: Wild-type, undomestica...   \n",
       "...                                                  ...   \n",
       "13646                                                NaN   \n",
       "13647                                                NaN   \n",
       "13648                                                NaN   \n",
       "13649                                                NaN   \n",
       "13650                                                NaN   \n",
       "\n",
       "                    GSM_organism_ch1  \n",
       "0         Corynebacterium glutamicum  \n",
       "1         Corynebacterium glutamicum  \n",
       "2         Corynebacterium glutamicum  \n",
       "3                  Bacillus subtilis  \n",
       "4                  Bacillus subtilis  \n",
       "...                              ...  \n",
       "13646  Streptococcus pyogenes M1 GAS  \n",
       "13647  Streptococcus pyogenes M1 GAS  \n",
       "13648  Streptococcus pyogenes M1 GAS  \n",
       "13649  Streptococcus pyogenes M1 GAS  \n",
       "13650  Streptococcus pyogenes M1 GAS  \n",
       "\n",
       "[7756 rows x 6 columns]"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_match = no_match[[\"TaxId\", \"GSM_taxid_ch1\", \"GPL_taxid\", \"GPL_organism\", \"GSM_characteristics_ch1\", \"GSM_organism_ch1\"]]\n",
    "no_match"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Corynebacterium glutamicum', 'Bacillus subtilis',\n",
       "       'Corynebacterium glutamicum ATCC 13032',\n",
       "       'Gluconobacter oxydans 621H', 'Escherichia coli BW25113',\n",
       "       'Escherichia coli str. K-12 substr. MG1655', 'Aspergillus niger',\n",
       "       'Allochromatium vinosum DSM 180', 'Escherichia coli',\n",
       "       'Escherichia coli APEC O2', 'Homo sapiens',\n",
       "       'Salmonella enterica subsp. enterica serovar Typhimurium str. 14028S',\n",
       "       'Escherichia coli K-12', 'Escherichia coli ATCC 25922',\n",
       "       'Escherichia coli BL21(DE3)',\n",
       "       'Escherichia coli O157:H7 str. Sakai', 'Escherichia coli O157:H7',\n",
       "       'Escherichia coli DH1',\n",
       "       'Escherichia coli str. K-12 substr. MG1655star',\n",
       "       'Escherichia coli W', 'Enterococcus faecalis',\n",
       "       'Enterococcus faecalis V583', 'Pseudomonas aeruginosa PAO1',\n",
       "       'Mycobacterium tuberculosis',\n",
       "       'Mycobacterium tuberculosis str. Erdman = ATCC 35801',\n",
       "       'Mycolicibacterium smegmatis', 'Macaca mulatta',\n",
       "       'Mycobacterium tuberculosis CDC1551',\n",
       "       'Mycolicibacterium smegmatis MC2 155',\n",
       "       'Mycobacterium avium subsp. paratuberculosis',\n",
       "       'Mycobacterium tuberculosis variant bovis BCG',\n",
       "       'Mycobacterium tuberculosis H37Ra', 'Pseudomonas aeruginosa',\n",
       "       'Pseudomonas aeruginosa PA1', 'Bruynoghevirus PaP3',\n",
       "       'Pseudomonas aeruginosa PAHM4', 'Staphylococcus aureus',\n",
       "       'Staphylococcus aureus subsp. aureus NCTC 8325',\n",
       "       'Staphylococcus epidermidis', 'Coxiella burnetii',\n",
       "       'Granulibacter bethesdensis', 'Chlamydia trachomatis',\n",
       "       'Rickettsia rickettsii', 'Streptomyces coelicolor',\n",
       "       'Streptomyces lividans', 'Streptococcus pneumoniae D39',\n",
       "       'Streptococcus pneumoniae R6', 'Streptococcus pneumoniae TIGR4',\n",
       "       'Streptococcus pseudopneumoniae ATCC BAA-960 = CCUG 49455',\n",
       "       'Streptococcus mitis', 'Streptococcus oralis',\n",
       "       'Streptococcus pyogenes M1 GAS', 'Borreliella burgdorferi',\n",
       "       'Yersinia pestis', 'Streptococcus pyogenes M49 591',\n",
       "       'Streptococcus pyogenes NZ131', 'Streptococcus pyogenes MGAS6180',\n",
       "       \"Streptococcus sp. 'group A'\",\n",
       "       'Streptococcus pyogenes;Macaca fascicularis'], dtype=object)"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "no_match.GSM_organism_ch1.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "rank_score = {\"strain\" : 0,\n",
    "              \"species\" : 1,\n",
    "              \"genus\" : 2,\n",
    "              \"family\" : 3,\n",
    "              \"order\" : 4,\n",
    "              \"class\" : 5,\n",
    "              \"phylum\" : 6,\n",
    "              \"superkingdom\" : 7,\n",
    "              \"no rank\" : 8}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "def is_same_specie(query_txid, subject_txid, taxonomy_relations):\n",
    "    try:\n",
    "        query_specie = taxonomy_relations[taxonomy_relations.TaxId == query_txid].reset_index().loc[0,\"TaxIdSpecie\"]\n",
    "        subject_specie = taxonomy_relations[taxonomy_relations.TaxId == subject_txid].reset_index().loc[0,\"TaxIdSpecie\"]\n",
    "        \n",
    "    except: return(False)\n",
    "\n",
    "    if  query_specie == subject_specie:\n",
    "        return(True)\n",
    "\n",
    "    return(False)\n",
    "\n",
    "def get_most_specific_txid(row, rel):\n",
    "    \"\"\"\n",
    "    Obtiene el txid más específico disponible para aquellos GSMs cuya taxonomy relation es ambigüa\n",
    "    Primero se obtienen los txids asociados a cada metadato, descartando aquellos que no pertenezcan \n",
    "    a la misma especie\n",
    "    Luego se obtiene el rango al que corresponden esos txids y se selecciona el de más baja jerarquía\n",
    "    \"\"\"\n",
    "    q_id = row.taxonID\n",
    "    gsm_id = {id for id in row.GSM_taxid_ch1.split(';') if is_same_specie(q_id, id, rel)}\n",
    "    print(gsm_id)\n",
    "    gpl_id = {id for id in row.GPL_taxid.split(';') if is_same_specie(q_id, id, rel)}\n",
    "    print(gpl_id)\n",
    "    gpl_org_id = {id.reset_index().loc[0, \"taxonID\"] for orgs in row.GPL_organism \n",
    "                  for org in orgs.split(';') \n",
    "                  if not (id:=rel[rel.ScientificName == org]).empty and is_same_specie(q_id, id,rel)}\n",
    "    print(gpl_org_id)\n",
    "    ch_id1 = {id.reset_index.loc().loc[0,\"taxonID\"] for orgs in row.GSM_organism_ch1 \n",
    "              for org in orgs.split(';') \n",
    "              if not (id:=rel[rel.ScientificName == org]).empty and is_same_specie(q_id,id,rel)}\n",
    "    print(ch_id1)\n",
    "    ch_id2 = set()\n",
    "    if type(chars := row.GSM_characteristics_ch1) != float: \n",
    "       for d in chars.split(';'):\n",
    "          if 'Strain:' in d or 'strain:' in d:\n",
    "             ch_id2 = {row.GSM_organism_ch1 + d.split(':')[1]}\n",
    "    \n",
    "    if ch_id2 and not (id:=rel[rel.ScientificNameSpecie == ch_id2]).empty:\n",
    "      ch_id2 = id.reset_index().loc[0,\"taxonID\"]\n",
    "    \n",
    "    print(ch_id2)\n",
    "    return()\n",
    "    all_ids = gsm_id | gpl_id | gpl_org_id | ch_id1  \n",
    "    all_ids = [(rank_score.get(rank, 8),id) for id in all_ids if id and (rank:=rel[rel.taxonID == id].reset_index().loc[0,\"Rank\"])]\n",
    "    return(sorted(all_ids)[0][1])\n",
    "    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "set()"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n",
    "{id for id in data.loc[10,\"GSM_taxid_ch1\"].split(';') if is_same_specie(data.loc[10,\"taxonID\"], id, rel)}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "set()\n",
      "set()\n",
      "set()\n",
      "set()\n",
      "{'Bacillus subtilis NDmed'}\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "()"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_most_specific_txid(data.iloc[3], rel)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "taxonID          224308\n",
       "GSM_taxid_ch1      1423\n",
       "Name: 3, dtype: object"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[\"GSM_or\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "bool(set())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 25 de septeimbre de 2023\n",
    "Icorporación del concepto al Código"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dependencias\n",
    "from Bio import Entrez\n",
    "import GEOparse\n",
    "import pandas as pd\n",
    "import re\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_taxonomy_data(id, max_tries=10):\n",
    "\n",
    "  try:\n",
    "    handle = Entrez.efetch(db='taxonomy', id=id, rettype='xml')\n",
    "    record = Entrez.read(handle)\n",
    "  except:\n",
    "    if max_tries:\n",
    "      return(get_taxonomy_data(id, max_tries-1))\n",
    "    else:\n",
    "      exit()\n",
    "\n",
    "  return(record[0])\n",
    "\n",
    "def get_txid_species(record):\n",
    "  if record['Rank'] == 'species':\n",
    "    return (record['TaxId'])\n",
    "  else:\n",
    "    lineag = [element['TaxId'] for element in record['LineageEx'] if element['Rank'] == 'species']\n",
    "    return(lineag[0])\n",
    "\n",
    "def get_taxonomy_relations(txidList):\n",
    "  relations = []\n",
    "  rank_big = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "  for txid in txidList:\n",
    "    taxonomy_data = get_taxonomy_data(txid)\n",
    "    row = {key:(value if key != 'LineageEx' else ';'.join([i['TaxId'] for i in value]))\n",
    "            for key,value in taxonomy_data.items()\n",
    "            if key in key in ['TaxId','ScientificName','Rank', 'ParentTaxId', 'LineageEx']}\n",
    "    txid_specie = get_txid_species(taxonomy_data)\n",
    "    specie_name = ' '.join(taxonomy_data['ScientificName'].split(' ')[0:2])\n",
    "    if not taxonomy_data['Rank'] in ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species']:\n",
    "      final_tag_txid = taxonomy_data['ScientificName'].split(' ')[-1]\n",
    "    else:\n",
    "      final_tag_txid = 'None'\n",
    "    row.update({'TaxIdSpecie': txid_specie, 'ScientificNameSpecie':specie_name, 'FinalTagName':final_tag_txid})\n",
    "    relations.append(row)\n",
    "  return(pd.DataFrame.from_dict(relations))\n",
    "\n",
    "def check_relation_txid(tx_query,tx_subject, metadata_taxo, characteristics):\n",
    "  rank_big = ['superkingdom', 'phylum', 'class', 'order', 'family', 'genus']\n",
    "  metadata_query = metadata_taxo[metadata_taxo['TaxId'] == str(tx_query)]\n",
    "  metadata_subject = metadata_taxo[metadata_taxo['TaxId'] == tx_subject]\n",
    "  if not list(metadata_query['Rank']) in rank_big and not list(metadata_subject['Rank']) in rank_big:\n",
    "    if list(metadata_query['Rank']) == list(metadata_subject['Rank']):\n",
    "      return('associated_same txid explicit')\n",
    "    else:\n",
    "      if list(metadata_query['TaxIdSpecie']) == list(metadata_subject['TaxIdSpecie']):\n",
    "        if type(characteristics) != float and list(metadata_query['FinalTagName'])[0] in characteristics:\n",
    "          return('associated_same txid implicit')\n",
    "        else:\n",
    "          return('not associated_ambiguous relationship')\n",
    "      else:\n",
    "        return('not associated_different specie')\n",
    "  else:\n",
    "    return('not associated_not prosecutable')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv(\"../230827FilterFunc/bigprueba/27082023Test.tsv\", sep=\"\\t\").rename(columns={\"taxonID\":\"TaxId\"})\n",
    "data.TaxId = data.TaxId.astype('str')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 147,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TaxId</th>\n",
       "      <th>GPL_organism</th>\n",
       "      <th>GPL_taxid</th>\n",
       "      <th>GSM_taxid_ch1</th>\n",
       "      <th>GSM_organism_ch1</th>\n",
       "      <th>GSM_characteristics_ch1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>224308</td>\n",
       "      <td>Corynebacterium glutamicum;Pseudomonas putida ...</td>\n",
       "      <td>1718;160488;224308;290633;511145</td>\n",
       "      <td>1718</td>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "      <td>genotype: ChrS-Ala245fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>224308</td>\n",
       "      <td>Corynebacterium glutamicum;Pseudomonas putida ...</td>\n",
       "      <td>1718;160488;224308;290633;511145</td>\n",
       "      <td>1718</td>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "      <td>genotype: ChrS-Ala245fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>224308</td>\n",
       "      <td>Corynebacterium glutamicum;Pseudomonas putida ...</td>\n",
       "      <td>1718;160488;224308;290633;511145</td>\n",
       "      <td>1718</td>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "      <td>genotype: ChrS-Ala245fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>224308</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>224308</td>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>224308</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "      <td>224308</td>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12295</th>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>1314</td>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12296</th>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>1314</td>\n",
       "      <td>160490</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12297</th>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>1314</td>\n",
       "      <td>160490</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12298</th>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>1314</td>\n",
       "      <td>160490</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12299</th>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>1314</td>\n",
       "      <td>160490</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12300 rows × 6 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        TaxId                                       GPL_organism  \\\n",
       "0      224308  Corynebacterium glutamicum;Pseudomonas putida ...   \n",
       "1      224308  Corynebacterium glutamicum;Pseudomonas putida ...   \n",
       "2      224308  Corynebacterium glutamicum;Pseudomonas putida ...   \n",
       "3      224308         Bacillus subtilis subsp. subtilis str. 168   \n",
       "4      224308         Bacillus subtilis subsp. subtilis str. 168   \n",
       "...       ...                                                ...   \n",
       "12295    1314                             Streptococcus pyogenes   \n",
       "12296    1314                             Streptococcus pyogenes   \n",
       "12297    1314                             Streptococcus pyogenes   \n",
       "12298    1314                             Streptococcus pyogenes   \n",
       "12299    1314                             Streptococcus pyogenes   \n",
       "\n",
       "                              GPL_taxid GSM_taxid_ch1  \\\n",
       "0      1718;160488;224308;290633;511145          1718   \n",
       "1      1718;160488;224308;290633;511145          1718   \n",
       "2      1718;160488;224308;290633;511145          1718   \n",
       "3                                224308          1423   \n",
       "4                                224308          1423   \n",
       "...                                 ...           ...   \n",
       "12295                              1314          1314   \n",
       "12296                              1314        160490   \n",
       "12297                              1314        160490   \n",
       "12298                              1314        160490   \n",
       "12299                              1314        160490   \n",
       "\n",
       "                    GSM_organism_ch1  \\\n",
       "0         Corynebacterium glutamicum   \n",
       "1         Corynebacterium glutamicum   \n",
       "2         Corynebacterium glutamicum   \n",
       "3                  Bacillus subtilis   \n",
       "4                  Bacillus subtilis   \n",
       "...                              ...   \n",
       "12295         Streptococcus pyogenes   \n",
       "12296  Streptococcus pyogenes M1 GAS   \n",
       "12297  Streptococcus pyogenes M1 GAS   \n",
       "12298  Streptococcus pyogenes M1 GAS   \n",
       "12299  Streptococcus pyogenes M1 GAS   \n",
       "\n",
       "                                 GSM_characteristics_ch1  \n",
       "0                                genotype: ChrS-Ala245fs  \n",
       "1                                genotype: ChrS-Ala245fs  \n",
       "2                                genotype: ChrS-Ala245fs  \n",
       "3      strain: NDmed;genotype: Wild-type, undomestica...  \n",
       "4      strain: NDmed;genotype: Wild-type, undomestica...  \n",
       "...                                                  ...  \n",
       "12295                                                NaN  \n",
       "12296                                                NaN  \n",
       "12297                                                NaN  \n",
       "12298                                                NaN  \n",
       "12299                                                NaN  \n",
       "\n",
       "[12300 rows x 6 columns]"
      ]
     },
     "execution_count": 147,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[[\"TaxId\", \"GPL_organism\", \"GPL_taxid\", \"GSM_taxid_ch1\", \"GSM_organism_ch1\",\"GSM_characteristics_ch1\"]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "metadata": {},
   "outputs": [],
   "source": [
    "gpl_ids = {id for ids in data.GPL_taxid for id in ids.split(';')}\n",
    "gsm_ids = {id for ids in data.GSM_taxid_ch1 for id in ids.split(';')}\n",
    "gpl_orgs = {id for ids in data.GPL_organism for id in ids.split(';')}\n",
    "gsm_orgs = {id for ids in data.GSM_organism_ch1 for id in ids.split(';')}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 149,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSM_organism_ch1</th>\n",
       "      <th>GSM_characteristics_ch1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "      <td>genotype: ChrS-Ala245fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "      <td>genotype: ChrS-Ala245fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "      <td>genotype: ChrS-Ala245fs</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12295</th>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12296</th>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12297</th>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12298</th>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12299</th>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>12300 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                    GSM_organism_ch1  \\\n",
       "0         Corynebacterium glutamicum   \n",
       "1         Corynebacterium glutamicum   \n",
       "2         Corynebacterium glutamicum   \n",
       "3                  Bacillus subtilis   \n",
       "4                  Bacillus subtilis   \n",
       "...                              ...   \n",
       "12295         Streptococcus pyogenes   \n",
       "12296  Streptococcus pyogenes M1 GAS   \n",
       "12297  Streptococcus pyogenes M1 GAS   \n",
       "12298  Streptococcus pyogenes M1 GAS   \n",
       "12299  Streptococcus pyogenes M1 GAS   \n",
       "\n",
       "                                 GSM_characteristics_ch1  \n",
       "0                                genotype: ChrS-Ala245fs  \n",
       "1                                genotype: ChrS-Ala245fs  \n",
       "2                                genotype: ChrS-Ala245fs  \n",
       "3      strain: NDmed;genotype: Wild-type, undomestica...  \n",
       "4      strain: NDmed;genotype: Wild-type, undomestica...  \n",
       "...                                                  ...  \n",
       "12295                                                NaN  \n",
       "12296                                                NaN  \n",
       "12297                                                NaN  \n",
       "12298                                                NaN  \n",
       "12299                                                NaN  \n",
       "\n",
       "[12300 rows x 2 columns]"
      ]
     },
     "execution_count": 149,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[[\"GSM_organism_ch1\", \"GSM_characteristics_ch1\"]]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 25 de septiembre de 2023"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dependencias\n",
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "Entrez.email = \"diegocar@lcg.unam.mx\"\n",
    "Entrez.api_key = \"34677fdcfd2f0659a7f9ee05ab6e44704f09\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rank(id):\n",
    "    try: \n",
    "       with Entrez.efetch(db=\"taxonomy\", id=id, retmode=\"xml\") as handle:\n",
    "          record = Entrez.read(handle)[0]\n",
    "    except:\n",
    "       return(get_rank(id))\n",
    "    return(record['Rank'])\n",
    "\n",
    "def get_lineage(id):\n",
    "    try:\n",
    "      with Entrez.efetch(db=\"taxonomy\", id=id, retmode=\"xml\") as handle:\n",
    "          record= Entrez.read(handle)[0]\n",
    "    except:\n",
    "       return(get_lineage(id))\n",
    "    return(record['LineageEx'])\n",
    "\n",
    "def get_taxonomy(txid, max=10):\n",
    "  try:\n",
    "    with Entrez.efetch(db='taxonomy', id=txid, rettype='xml') as handle:\n",
    "      record = Entrez.read(handle)[0]\n",
    "  except:\n",
    "     if max: return(get_taxonomy(id, max-1))\n",
    "     print(\"Estoy cansado jefe...\")\n",
    "     exit()\n",
    "  \n",
    "  rank = record['LineageEx'][-1]['Rank']\n",
    "  org = record['ScientificName']\n",
    "  name = record['ScientificName']\n",
    "  for d in record['LineageEx']:\n",
    "     if d.get(\"Rank\") == \"species\":\n",
    "        org = d.get(\"ScientificName\")\n",
    "  return((rank,org,name))\n",
    "\n",
    "def get_scientific_name(id):\n",
    "   try:\n",
    "      with Entrez.efetch(db=\"taxonomy\", id=id, retmode=\"xml\") as handle:\n",
    "         record = Entrez.read(handle)[0]\n",
    "   except:\n",
    "       return(get_scientific_name(id))\n",
    "   return(record['ScientificName'])\n",
    "\n",
    "def is_same_specie(q_id, s_id, tax_rel):\n",
    "  q_specie = tax_rel[tax_rel.TaxId == q_id].reset_index().loc[0,\"Specie\"]\n",
    "  s_specie = tax_rel[tax_rel.TaxId == s_id].reset_index().loc[0,\"Specie\"]\n",
    "  if q_specie == s_specie: return(True)\n",
    "  return(False)\n",
    "\n",
    "def get_best_id(metadata, tax_rel):\n",
    "   ch1_id = metadata.GSM_taxid_ch1\n",
    "   q_id = metadata.TaxId\n",
    "   if not is_same_specie(q_id, ch1_id, tax_rel): return(ch1_id)\n",
    "\n",
    "   ch1_id = {ch1_id}\n",
    "   q_id = metadata.TaxId\n",
    "   gpl_ids = {id for id in metadata.GPL_taxid.split(';') if is_same_specie(q_id, id, tax_rel)}\n",
    "   gsm_ids = {id for id in metadata.GSM_taxid_ch1.split(';') if is_same_specie(q_id, id, tax_rel)}\n",
    "   \n",
    "\n",
    "   ids = gpl_ids | gsm_ids | ch1_id\n",
    "\n",
    "   rank_score = {\"subspecies\": 0,\n",
    "                 \"strain\" : 0,\n",
    "                 \"serotype\": 0,\n",
    "                 \"species group\": 1,\n",
    "                 \"species\" : 1,\n",
    "                 \"genus\" : 2,\n",
    "                 \"family\" : 3,\n",
    "                 \"order\" : 4,\n",
    "                 \"class\" : 5,\n",
    "                 \"phylum\" : 6,\n",
    "                 \"superkingdom\" : 7,\n",
    "                 \"no rank\" : 8}\n",
    "\n",
    "   rank = [(rank_score.get(tax_rel[tax_rel.TaxId == id].reset_index().loc[0,\"Ranks\"], 8), id) for id in ids]\n",
    "   return(sorted(rank)[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12300, 25)"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"../230827FilterFunc/bigprueba/27082023Test.tsv\", sep=\"\\t\").rename(columns={\"taxonID\":\"TaxId\"})\n",
    "data.TaxId = data.TaxId.astype(\"str\")\n",
    "rela = pd.read_csv(\"./taxonomy_relations_mini.tsv\", sep='\\t')\n",
    "rela.TaxId = rela.TaxId.astype(\"str\")\n",
    "data.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "txid_list = set(data['TaxId'])\n",
    "txid_ch1 = {id for ids in data.GSM_taxid_ch1 for id in ids.split(';')}\n",
    "txid_gpl = {id for ids in data.GPL_taxid for id in ids.split(';')}\n",
    "txid_all = txid_list | txid_gpl | txid_ch1\n",
    "txids,ranks,names,sn = zip(*[(id, *get_taxonomy(id)) for id in  txid_all])\n",
    "info = pd.DataFrame.from_dict({\"TaxId\" : txids, \"Ranks\" : ranks, \"Specie\" : names, \"ScientificName\" : sn})\n",
    "info.to_csv(\"relations.tsv\", sep='\\t', index=False, header=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'TaxId': '131567', 'ScientificName': 'cellular organisms', 'Rank': 'no rank'}\n",
      "{'TaxId': '2', 'ScientificName': 'Bacteria', 'Rank': 'superkingdom'}\n",
      "{'TaxId': '1783272', 'ScientificName': 'Terrabacteria group', 'Rank': 'clade'}\n",
      "{'TaxId': '1239', 'ScientificName': 'Bacillota', 'Rank': 'phylum'}\n",
      "{'TaxId': '91061', 'ScientificName': 'Bacilli', 'Rank': 'class'}\n",
      "{'TaxId': '1385', 'ScientificName': 'Bacillales', 'Rank': 'order'}\n",
      "{'TaxId': '186817', 'ScientificName': 'Bacillaceae', 'Rank': 'family'}\n",
      "{'TaxId': '1386', 'ScientificName': 'Bacillus', 'Rank': 'genus'}\n",
      "{'TaxId': '653685', 'ScientificName': 'Bacillus subtilis group', 'Rank': 'species group'}\n"
     ]
    }
   ],
   "source": [
    "with Entrez.efetch(db=\"taxonomy\", id=1423, retmode=\"xml\") as handle:\n",
    "    record = Entrez.read(handle)[0]\n",
    "for d in record['LineageEx']:\n",
    "    print(d)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'1280'"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_best_id(data.iloc[12000], info)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TaxId</th>\n",
       "      <th>Ranks</th>\n",
       "      <th>Specie</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>1423</td>\n",
       "      <td>species group</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   TaxId          Ranks             Specie     ScientificName\n",
       "18  1423  species group  Bacillus subtilis  Bacillus subtilis"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info[info.TaxId == \"1423\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>TaxId</th>\n",
       "      <th>Ranks</th>\n",
       "      <th>Specie</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>224308</td>\n",
       "      <td>subspecies</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>Bacillus subtilis subsp. subtilis str. 168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>67</th>\n",
       "      <td>1718</td>\n",
       "      <td>genus</td>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "      <td>Corynebacterium glutamicum</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     TaxId       Ranks                      Specie  \\\n",
       "16  224308  subspecies           Bacillus subtilis   \n",
       "67    1718       genus  Corynebacterium glutamicum   \n",
       "\n",
       "                                ScientificName  \n",
       "16  Bacillus subtilis subsp. subtilis str. 168  \n",
       "67                  Corynebacterium glutamicum  "
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "info[info.TaxId.isin([\"1718\", \"224308\"])]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# 27 de septiembre de 2023\n",
    "Resolución de ambigüedades taxonómicas\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Dependencias\n",
    "from Bio import Entrez\n",
    "import pandas as pd\n",
    "Entrez.email = \"diegocar@lcg.unam.mx\"\n",
    "Entrez.api_key = \"34677fdcfd2f0659a7f9ee05ab6e44704f09\"\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 143,
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_rank(id):\n",
    "    try: \n",
    "       with Entrez.efetch(db=\"taxonomy\", id=id, retmode=\"xml\") as handle:\n",
    "          record = Entrez.read(handle)[0]\n",
    "    except:\n",
    "       return(get_rank(id))\n",
    "    return(record['Rank'])\n",
    "\n",
    "def get_lineage(id):\n",
    "    try:\n",
    "      with Entrez.efetch(db=\"taxonomy\", id=id, retmode=\"xml\") as handle:\n",
    "          record= Entrez.read(handle)[0]\n",
    "    except:\n",
    "       return(get_lineage(id))\n",
    "    return(record['LineageEx'])\n",
    "\n",
    "def get_taxonomy(txid, max=15):\n",
    "  try:\n",
    "    with Entrez.efetch(db='taxonomy', id=txid, rettype='xml') as handle:\n",
    "      record = Entrez.read(handle)[0]\n",
    "  except:\n",
    "     if max: return(get_taxonomy(txid, max-1))\n",
    "     print(f\"Request error in getting {txid} taxonomy\")\n",
    "     return((\"\",\"\",\"\"))\n",
    "  \n",
    "  rank = record['LineageEx'][-1]['Rank']\n",
    "  org = record['ScientificName']\n",
    "  name = record['ScientificName']\n",
    "  for d in record['LineageEx']:\n",
    "     if d.get(\"Rank\") == \"species\":\n",
    "        org = d.get(\"ScientificName\")\n",
    "  return((rank,org,name))\n",
    "\n",
    "def get_scientific_name(id):\n",
    "   try:\n",
    "      with Entrez.efetch(db=\"taxonomy\", id=id, retmode=\"xml\") as handle:\n",
    "         record = Entrez.read(handle)[0]\n",
    "   except:\n",
    "       return(get_scientific_name(id))\n",
    "   return(record['ScientificName'])\n",
    "\n",
    "def is_same_specie(q_id, s_id, tax_rel):\n",
    "   if (q_specie := tax_rel[tax_rel.TaxId == q_id]).empty or (s_specie := tax_rel[tax_rel.TaxId == s_id]).empty:\n",
    "      print(f\"Can´t compare {q_id} with {s_id}\")\n",
    "      return(False)\n",
    "\n",
    "   q_specie = q_specie.reset_index().loc[0,\"Specie\"]\n",
    "   s_specie = s_specie.reset_index().loc[0,\"Specie\"]\n",
    "\n",
    "   if q_specie == s_specie: return(True)\n",
    "   return(False)\n",
    "\n",
    "def get_best_id(metadata, tax_rel):\n",
    "   ch1_id = metadata.GSM_taxid_ch1\n",
    "   q_id = metadata.TaxId\n",
    "   if not is_same_specie(q_id, ch1_id, tax_rel): return(ch1_id)\n",
    "\n",
    "   ch1_id = {ch1_id}\n",
    "   gpl_ids = {id for id in metadata.GPL_taxid.split(';') if is_same_specie(q_id, id, tax_rel)} \n",
    "   gsm_ids = {id for id in metadata.GSM_taxid_ch1.split(';') if is_same_specie(q_id, id, tax_rel)} \n",
    "   \n",
    "\n",
    "   ids = gpl_ids | gsm_ids | ch1_id\n",
    "\n",
    "   rank_score = {\"subspecies\": 0,\n",
    "                 \"strain\" : 0,\n",
    "                 \"serotype\": 0,\n",
    "                 \"species group\": 1,\n",
    "                 \"species\" : 1,\n",
    "                 \"genus\" : 2,\n",
    "                 \"family\" : 3,\n",
    "                 \"order\" : 4,\n",
    "                 \"class\" : 5,\n",
    "                 \"phylum\" : 6,\n",
    "                 \"superkingdom\" : 7,\n",
    "                 \"no rank\" : 8}\n",
    "\n",
    "   rank = [(rank_score.get(tax_rel[tax_rel.TaxId == id].reset_index().loc[0,\"Rank\"], 8), id) for id in ids]\n",
    "   return(sorted(rank)[0][1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 144,
   "metadata": {},
   "outputs": [],
   "source": [
    "def apply_GEO_filters(raw_metadata, rank):\n",
    "    # Delete superseries\n",
    "    super_GSEs = [index for index,row in raw_metadata.iterrows() if 'SuperSeries' in row['GSE_relation']]\n",
    "    filtered = raw_metadata.drop(super_GSEs).reset_index().drop('index', 1)\n",
    "    print(\"Deleted SuperSeries\")\n",
    "\n",
    "    # Delete tech != array\n",
    "    interest_techs = ['in situ oligonucleotide', 'spotted oligonucleotide', 'spotted DNA/cDNA', 'mixed spotted oligonucleotide/cDNA']\n",
    "    not_array = [index for index,row in filtered.iterrows() if row['GPL_technology'] not in interest_techs]\n",
    "    filtered = filtered.drop(not_array).reset_index().drop('index', 1)\n",
    "    print(\"Deleted !array tech\")\n",
    "\n",
    "    # Delete spurious txIDs\n",
    "    txid_query = set(filtered['TaxId'])\n",
    "    txid_ch1 = {id for ids in filtered.GSM_taxid_ch1 for id in ids.split(';')} #############\n",
    "    txid_gpl = {id for ids in filtered.GPL_taxid for id in ids.split(';')} #####################\n",
    "    txid_all = txid_query | txid_gpl | txid_ch1\n",
    "    txids,ranks,specie,name = zip(*[(id, *get_taxonomy(id)) for id in  txid_all])\n",
    "    tax_rel = pd.DataFrame.from_dict({\"TaxId\":txids, \"Rank\":ranks, \"Specie\":specie, \"ScientificName\":name})\n",
    "    tax_rel.TaxId = tax_rel.TaxId.astype(\"str\")\n",
    "    tax_rel.to_csv(f'taxonomyRelationsFunction2.tsv', sep='\\t')\n",
    "    #tax_rel = pd.read_csv(\"./taxonomyRelationsFunction2.tsv\", sep='\\t')\n",
    "    #tax_rel.TaxId = tax_rel.TaxId.astype(\"str\")\n",
    "    print(\"Tax rel constructed\")\n",
    "    \n",
    "    if rank == \"specie\":\n",
    "      spurious = [i for i,r in filtered.iterrows() if not is_same_specie(r.TaxId, r.GSM_taxid_ch1,tax_rel)]\n",
    "      filtered = filtered.drop(spurious).reset_index().drop('index', 1)\n",
    "      print(\"Spurious GSMs deleted\")\n",
    "\n",
    "    elif rank == \"sameTaxId\":\n",
    "      sameTaxid = [i for i,r in filtered.iterrows() if r.TaxId == r.GSM_taxid_ch1]\n",
    "      taxid_expl = filtered.iloc[sameTaxid]\n",
    "      print(\"Explicit txid relation GSM saved\")\n",
    "      filtered = filtered.drop(sameTaxid)\n",
    "\n",
    "      recovered = [i for i,r in filtered.iterrows() if r.TaxId == get_best_id(r, tax_rel)]\n",
    "      print(\"Recoverable  GSMs identified\")\n",
    "      taxid_rec = filtered.loc[recovered]\n",
    "\n",
    "      filtered = pd.concat([taxid_expl, taxid_rec]).reset_index().drop('index', 1)\n",
    "      \n",
    "    return(filtered)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(12300, 25)"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data = pd.read_csv(\"../230827FilterFunc/bigprueba/27082023Test.tsv\", sep=\"\\t\").rename(columns={\"taxonID\":\"TaxId\"})\n",
    "data.TaxId = data.TaxId.astype(\"str\")\n",
    "rela = pd.read_csv(\"./taxonomy_relations_mini.tsv\", sep='\\t')\n",
    "rela.TaxId = rela.TaxId.astype(\"str\")\n",
    "data.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prueba para \"specie\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 146,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Deleted SuperSeries\n",
      "Deleted !array tech\n",
      "Tax rel constructed\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Can´t compare 1314 with 1314;9541\n",
      "Spurious GSMs deleted\n"
     ]
    }
   ],
   "source": [
    "filtered = apply_GEO_filters(data, \"specie\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "rel = pd.read_csv(\"./taxonomyRelationsFunction.tsv\", sep='\\t')\n",
    "rel.TaxId = rel.TaxId.astype(\"str\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "mega = pd.merge(rel, filtered)[[\"TaxId\", \"ScientificName\", \"Specie\", \"GSM_taxid_ch1\"]]\n",
    "not_identical_taxids = {row.GSM_taxid_ch1 for i,row in mega.iterrows() if row.TaxId != row.GSM_taxid_ch1}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "29"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "txid,gsm_species = zip(*[(id, get_scientific_name(id)) for id in not_identical_taxids])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "mega.merge(pd.DataFrame.from_dict({\"GSM_taxid_ch1\":txid, \"GSM_specie_ch1\":gsm_species})).to_csv(\"notIdenticalTaxIDs.tsv\", sep='\\t')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prueba 1 para \"sameTaxid\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "apply_GEO_filters(data, \"sameTaxId\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Can´t make txids 511145 588858 comparation\n",
      "Can´t make txids 511145 588858 comparation\n",
      "Can´t make txids 511145 588858 comparation\n",
      "Can´t make txids 511145 588858 comparation\n",
      "Can´t make txids 511145 588858 comparation\n",
      "Can´t make txids 511145 588858 comparation\n",
      "Can´t make txids 511145 588858 comparation\n",
      "Can´t make txids 511145 588858 comparation\n",
      "Can´t make txids 511145 83334 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 511145 879462 comparation\n",
      "Can´t make txids 199310 588858 comparation\n",
      "Can´t make txids 199310 588858 comparation\n",
      "Can´t make txids 199310 588858 comparation\n",
      "Can´t make txids 199310 588858 comparation\n",
      "Can´t make txids 199310 588858 comparation\n",
      "Can´t make txids 199310 588858 comparation\n",
      "Can´t make txids 199310 588858 comparation\n",
      "Can´t make txids 199310 588858 comparation\n",
      "Can´t make txids 83332 9544 comparation\n",
      "Can´t make txids 83332 9544 comparation\n",
      "Can´t make txids 83332 9544 comparation\n",
      "Can´t make txids 83332 9544 comparation\n",
      "Can´t make txids 83332 9544 comparation\n",
      "Can´t make txids 83332 9544 comparation\n",
      "Can´t make txids 83332 9544 comparation\n",
      "Can´t make txids 83332 9544 comparation\n",
      "Can´t make txids 83332 1770 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n"
     ]
    }
   ],
   "source": [
    "tax_rel = pd.read_csv(\"./taxonomyRelationsFunction2.tsv\", sep=\"\\t\")\n",
    "tax_rel.TaxId = tax_rel.TaxId.astype(\"str\")\n",
    "sameTaxid = [i for i,r in data.iterrows() if r.TaxId == r.GSM_taxid_ch1]\n",
    "taxid_expl = data.iloc[sameTaxid]\n",
    "fltrd = data.drop(sameTaxid)\n",
    "\n",
    "recovered = [i for i,r in fltrd.iterrows() if r.TaxId == get_best_id(r, tax_rel)]\n",
    "taxid_rec = fltrd.loc[recovered]\n",
    "\n",
    "fltrd = pd.concat([taxid_expl, taxid_rec]).reset_index().drop('index', 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GPL_taxid</th>\n",
       "      <th>GSM_taxid_ch1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TaxId</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>224308</th>\n",
       "      <td>1423</td>\n",
       "      <td>224308</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196627</th>\n",
       "      <td>442;562;1718;160488;224308</td>\n",
       "      <td>196627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196627</th>\n",
       "      <td>1718</td>\n",
       "      <td>196627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196627</th>\n",
       "      <td>442;562;1718;224308</td>\n",
       "      <td>196627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196627</th>\n",
       "      <td>303;442;562;1718;224308</td>\n",
       "      <td>196627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511145</th>\n",
       "      <td>83333</td>\n",
       "      <td>511145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511145</th>\n",
       "      <td>442;562;1718;224308</td>\n",
       "      <td>511145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511145</th>\n",
       "      <td>562</td>\n",
       "      <td>511145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511145</th>\n",
       "      <td>155864;199310;386585;511145</td>\n",
       "      <td>511145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511145</th>\n",
       "      <td>1392;99287;155864;199310;226185;226186;386585;...</td>\n",
       "      <td>511145</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>364106</th>\n",
       "      <td>562</td>\n",
       "      <td>364106</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>331111</th>\n",
       "      <td>562;620</td>\n",
       "      <td>331111</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316385</th>\n",
       "      <td>562</td>\n",
       "      <td>316385</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199310</th>\n",
       "      <td>562</td>\n",
       "      <td>199310</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83332</th>\n",
       "      <td>1765;83332</td>\n",
       "      <td>83332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83332</th>\n",
       "      <td>1773</td>\n",
       "      <td>83332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83332</th>\n",
       "      <td>83331;83332</td>\n",
       "      <td>83332</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208964</th>\n",
       "      <td>287</td>\n",
       "      <td>208964</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>426430</th>\n",
       "      <td>1280</td>\n",
       "      <td>426430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158879</th>\n",
       "      <td>1280</td>\n",
       "      <td>158879</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158878</th>\n",
       "      <td>1280</td>\n",
       "      <td>158878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93062</th>\n",
       "      <td>1280</td>\n",
       "      <td>93062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100226</th>\n",
       "      <td>1902</td>\n",
       "      <td>100226</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1313</th>\n",
       "      <td>373153</td>\n",
       "      <td>1313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1313</th>\n",
       "      <td>171101</td>\n",
       "      <td>1313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1313</th>\n",
       "      <td>170187</td>\n",
       "      <td>1313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1313</th>\n",
       "      <td>170187;171101;512566</td>\n",
       "      <td>1313</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1314</th>\n",
       "      <td>471876</td>\n",
       "      <td>1314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1314</th>\n",
       "      <td>139;632;777;813;1280;1314</td>\n",
       "      <td>1314</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224308</th>\n",
       "      <td>224308</td>\n",
       "      <td>1423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>224308</th>\n",
       "      <td>562;224308</td>\n",
       "      <td>1423</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196627</th>\n",
       "      <td>196627</td>\n",
       "      <td>1718</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511145</th>\n",
       "      <td>511145</td>\n",
       "      <td>562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511145</th>\n",
       "      <td>511145</td>\n",
       "      <td>536056</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511145</th>\n",
       "      <td>511145</td>\n",
       "      <td>566546</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>511145</th>\n",
       "      <td>511145</td>\n",
       "      <td>83333</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>199310</th>\n",
       "      <td>199310</td>\n",
       "      <td>562</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>83332</th>\n",
       "      <td>83332</td>\n",
       "      <td>33892</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>208964</th>\n",
       "      <td>208964</td>\n",
       "      <td>287</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196620</th>\n",
       "      <td>196620</td>\n",
       "      <td>1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>196620</th>\n",
       "      <td>777;783;83560;115711;176279;176280;196620;2243...</td>\n",
       "      <td>1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158879</th>\n",
       "      <td>158879</td>\n",
       "      <td>1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158879</th>\n",
       "      <td>158879</td>\n",
       "      <td>93061</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158878</th>\n",
       "      <td>1280;93061;93062;158878;158879;196620;282458;2...</td>\n",
       "      <td>1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93062</th>\n",
       "      <td>93062</td>\n",
       "      <td>1280</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>100226</th>\n",
       "      <td>100226</td>\n",
       "      <td>1902</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                GPL_taxid GSM_taxid_ch1\n",
       "TaxId                                                                  \n",
       "224308                                               1423        224308\n",
       "196627                         442;562;1718;160488;224308        196627\n",
       "196627                                               1718        196627\n",
       "196627                                442;562;1718;224308        196627\n",
       "196627                            303;442;562;1718;224308        196627\n",
       "511145                                              83333        511145\n",
       "511145                                442;562;1718;224308        511145\n",
       "511145                                                562        511145\n",
       "511145                        155864;199310;386585;511145        511145\n",
       "511145  1392;99287;155864;199310;226185;226186;386585;...        511145\n",
       "364106                                                562        364106\n",
       "331111                                            562;620        331111\n",
       "316385                                                562        316385\n",
       "199310                                                562        199310\n",
       "83332                                          1765;83332         83332\n",
       "83332                                                1773         83332\n",
       "83332                                         83331;83332         83332\n",
       "208964                                                287        208964\n",
       "426430                                               1280        426430\n",
       "158879                                               1280        158879\n",
       "158878                                               1280        158878\n",
       "93062                                                1280         93062\n",
       "100226                                               1902        100226\n",
       "1313                                               373153          1313\n",
       "1313                                               171101          1313\n",
       "1313                                               170187          1313\n",
       "1313                                 170187;171101;512566          1313\n",
       "1314                                               471876          1314\n",
       "1314                            139;632;777;813;1280;1314          1314\n",
       "224308                                             224308          1423\n",
       "224308                                         562;224308          1423\n",
       "196627                                             196627          1718\n",
       "511145                                             511145           562\n",
       "511145                                             511145        536056\n",
       "511145                                             511145        566546\n",
       "511145                                             511145         83333\n",
       "199310                                             199310           562\n",
       "83332                                               83332         33892\n",
       "208964                                             208964           287\n",
       "196620                                             196620          1280\n",
       "196620  777;783;83560;115711;176279;176280;196620;2243...          1280\n",
       "158879                                             158879          1280\n",
       "158879                                             158879         93061\n",
       "158878  1280;93061;93062;158878;158879;196620;282458;2...          1280\n",
       "93062                                               93062          1280\n",
       "100226                                             100226          1902"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fltrd.loc[[i for i,r in fltrd.iterrows() if r.GPL_taxid != r.GSM_taxid_ch1], [\"TaxId\",\"GPL_taxid\", \"GSM_taxid_ch1\"]].set_index(\"TaxId\").drop_duplicates()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Unnamed: 0</th>\n",
       "      <th>TaxId</th>\n",
       "      <th>Rank</th>\n",
       "      <th>Specie</th>\n",
       "      <th>ScientificName</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: [Unnamed: 0, TaxId, Rank, Specie, ScientificName]\n",
       "Index: []"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "tax_rel[tax_rel.TaxId == \"588858\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'Salmonella enterica subsp. enterica serovar Typhimurium str. 14028S'"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "get_scientific_name(588858)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### WTF\n",
    "Algunos txids no fueron comparables (archivo txidfilter.txt)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "wtf = {id : get_scientific_name(id) for id in [\"588858\", \"83334\", \"879462\", \"588858\", \"9544\", \"1770\", \"1314;9541\"]}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "wtfks = {\"588858\", \"83334\", \"879462\", \"588858\", \"9544\", \"1770\", \"1314;9541\"}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['588858', '83334', '879462', '588858', '9544', '1770', '1314;9541']"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "wtfks"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [],
   "source": [
    "filter_set = {id for ids in data.GSM_taxid_ch1 for id in ids.split(';')}\n",
    "best_set = {id for i,r in data.iterrows() for id in r.GSM_taxid_ch1.split(\";\")}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "76\n",
      "76\n",
      "76\n"
     ]
    }
   ],
   "source": [
    "print(len(filter_set))\n",
    "print(len(best_set))\n",
    "print(len(best_set & filter_set))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "86\n",
      "86\n",
      "86\n"
     ]
    }
   ],
   "source": [
    "filter_set = {id for ids in data.GPL_taxid for id in ids.split(';')}\n",
    "best_set = {id for i,r in data.iterrows() for id in r.GPL_taxid.split(\";\")}\n",
    "print(len(filter_set))\n",
    "print(len(best_set))\n",
    "print(len(best_set & filter_set))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "('strain', 'Salmonella enterica', 'Salmonella enterica subsp. enterica serovar Typhimurium str. 14028S')\n",
      "('species', 'Escherichia coli', 'Escherichia coli O157:H7')\n",
      "('strain', 'Escherichia coli', 'Escherichia coli str. K-12 substr. MG1655star')\n",
      "('strain', 'Salmonella enterica', 'Salmonella enterica subsp. enterica serovar Typhimurium str. 14028S')\n",
      "('genus', 'Macaca mulatta', 'Macaca mulatta')\n",
      "('species', 'Mycobacterium avium', 'Mycobacterium avium subsp. paratuberculosis')\n",
      "('genus', 'Streptococcus pyogenes', 'Streptococcus pyogenes')\n"
     ]
    }
   ],
   "source": [
    "for i in [get_taxonomy(id) for id in wtfks]:\n",
    "    print(i)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [],
   "source": [
    "gsm_wtfks={id for i,row in data.iterrows() for id in row.GSM_taxid_ch1.split(';') if id in wtfks}\n",
    "gpl_wtfks={id for i,row in data.iterrows() for id in row.GPL_taxid.split(';') if id in wtfks}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'9544', '1314;9541', '879462', '83334', '588858', '1770'}\n",
      "{'9544', '879462', '83334', '588858', '1770'}\n",
      "{'588858', '879462', '9544'}\n"
     ]
    }
   ],
   "source": [
    "print(wtfks)\n",
    "print(gsm_wtfks)\n",
    "print(gpl_wtfks)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Prueba 2 para \"sameTaxid\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 133,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Deleted SuperSeries\n",
      "Deleted !array tech\n",
      "Request error in getting <built-in function id> taxonomy\n",
      "Request error in getting <built-in function id> taxonomy\n",
      "Request error in getting <built-in function id> taxonomy\n",
      "Tax rel constructed\n",
      "Explicit txid relation GSM saved\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Can´t make txids 1314 1314;9541 comparation\n",
      "Recoverable  GSMs identified\n"
     ]
    }
   ],
   "source": [
    "# Filtrado para quedarse con tax Ids idénticos\n",
    "filtrado = apply_GEO_filters(data, \"sameTaxId\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 151,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>GSE_title</th>\n",
       "      <th>GSE_geo_accession</th>\n",
       "      <th>GSE_type</th>\n",
       "      <th>GSE_relation</th>\n",
       "      <th>TaxId</th>\n",
       "      <th>GSE_supplementary_file</th>\n",
       "      <th>GPL_title</th>\n",
       "      <th>GPL_geo_accession</th>\n",
       "      <th>GPL_technology</th>\n",
       "      <th>GPL_distribution</th>\n",
       "      <th>...</th>\n",
       "      <th>GSM_type</th>\n",
       "      <th>GSM_geo_accession</th>\n",
       "      <th>GSM_channel_count</th>\n",
       "      <th>GSM_source_name_ch1</th>\n",
       "      <th>GSM_taxid_ch1</th>\n",
       "      <th>GSM_organism_ch1</th>\n",
       "      <th>GSM_characteristics_ch1</th>\n",
       "      <th>GSM_taxid_ch2</th>\n",
       "      <th>GSM_characteristics_ch2</th>\n",
       "      <th>GSM_supplementary_file</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Temporal transcriptome analysis of Bacillus su...</td>\n",
       "      <td>GSE190460</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>224308</td>\n",
       "      <td>geo/series/GSE190nnn/GSE190460/suppl/</td>\n",
       "      <td>BaSysBio Bacillus subtilis T3 400K array</td>\n",
       "      <td>GPL21981</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>custom-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM5724096</td>\n",
       "      <td>1</td>\n",
       "      <td>Bacillus subtilis NDmed</td>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>geo/samples/GSM5724nnn/GSM5724096/suppl/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Temporal transcriptome analysis of Bacillus su...</td>\n",
       "      <td>GSE190460</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>224308</td>\n",
       "      <td>geo/series/GSE190nnn/GSE190460/suppl/</td>\n",
       "      <td>BaSysBio Bacillus subtilis T3 400K array</td>\n",
       "      <td>GPL21981</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>custom-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM5724097</td>\n",
       "      <td>1</td>\n",
       "      <td>Bacillus subtilis NDmed</td>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>geo/samples/GSM5724nnn/GSM5724097/suppl/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Temporal transcriptome analysis of Bacillus su...</td>\n",
       "      <td>GSE190460</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>224308</td>\n",
       "      <td>geo/series/GSE190nnn/GSE190460/suppl/</td>\n",
       "      <td>BaSysBio Bacillus subtilis T3 400K array</td>\n",
       "      <td>GPL21981</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>custom-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM5724098</td>\n",
       "      <td>1</td>\n",
       "      <td>Bacillus subtilis NDmed</td>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>geo/samples/GSM5724nnn/GSM5724098/suppl/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Temporal transcriptome analysis of Bacillus su...</td>\n",
       "      <td>GSE190460</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>224308</td>\n",
       "      <td>geo/series/GSE190nnn/GSE190460/suppl/</td>\n",
       "      <td>BaSysBio Bacillus subtilis T3 400K array</td>\n",
       "      <td>GPL21981</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>custom-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM5724099</td>\n",
       "      <td>1</td>\n",
       "      <td>Bacillus subtilis NDmed</td>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>geo/samples/GSM5724nnn/GSM5724099/suppl/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Temporal transcriptome analysis of Bacillus su...</td>\n",
       "      <td>GSE190460</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>224308</td>\n",
       "      <td>geo/series/GSE190nnn/GSE190460/suppl/</td>\n",
       "      <td>BaSysBio Bacillus subtilis T3 400K array</td>\n",
       "      <td>GPL21981</td>\n",
       "      <td>in situ oligonucleotide</td>\n",
       "      <td>custom-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM5724100</td>\n",
       "      <td>1</td>\n",
       "      <td>Bacillus subtilis NDmed</td>\n",
       "      <td>1423</td>\n",
       "      <td>Bacillus subtilis</td>\n",
       "      <td>strain: NDmed;genotype: Wild-type, undomestica...</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>geo/samples/GSM5724nnn/GSM5724100/suppl/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8890</th>\n",
       "      <td>M1 and M6 serotypes, Wild-type vs. Mga- GAS</td>\n",
       "      <td>GSE1895</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>1314</td>\n",
       "      <td>geo/series/GSE1nnn/GSE1895/suppl/</td>\n",
       "      <td>M1, M3, M18 GAS Array</td>\n",
       "      <td>GPL1482</td>\n",
       "      <td>spotted oligonucleotide</td>\n",
       "      <td>non-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM33613</td>\n",
       "      <td>2</td>\n",
       "      <td>CsCl total RNA from M1 serotype GAS KSM165L</td>\n",
       "      <td>1314</td>\n",
       "      <td>Streptococcus pyogenes</td>\n",
       "      <td>NaN</td>\n",
       "      <td>1314.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>geo/samples/GSM33nnn/GSM33613/suppl/</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8891</th>\n",
       "      <td>Wild-type and irr mutant Group A Streptococcus...</td>\n",
       "      <td>GSE1550</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>1314</td>\n",
       "      <td>None</td>\n",
       "      <td>NIAID Group A Streptococcus</td>\n",
       "      <td>GPL1338</td>\n",
       "      <td>spotted DNA/cDNA</td>\n",
       "      <td>non-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM26678</td>\n",
       "      <td>2</td>\n",
       "      <td>wt JRS4</td>\n",
       "      <td>160490</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160490.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8892</th>\n",
       "      <td>Wild-type and irr mutant Group A Streptococcus...</td>\n",
       "      <td>GSE1550</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>1314</td>\n",
       "      <td>None</td>\n",
       "      <td>NIAID Group A Streptococcus</td>\n",
       "      <td>GPL1338</td>\n",
       "      <td>spotted DNA/cDNA</td>\n",
       "      <td>non-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM26679</td>\n",
       "      <td>2</td>\n",
       "      <td>wt JRS4</td>\n",
       "      <td>160490</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160490.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8893</th>\n",
       "      <td>Wild-type and irr mutant Group A Streptococcus...</td>\n",
       "      <td>GSE1550</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>1314</td>\n",
       "      <td>None</td>\n",
       "      <td>NIAID Group A Streptococcus</td>\n",
       "      <td>GPL1338</td>\n",
       "      <td>spotted DNA/cDNA</td>\n",
       "      <td>non-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM26680</td>\n",
       "      <td>2</td>\n",
       "      <td>wt JRS4</td>\n",
       "      <td>160490</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160490.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8894</th>\n",
       "      <td>Wild-type and irr mutant Group A Streptococcus...</td>\n",
       "      <td>GSE1550</td>\n",
       "      <td>Expression profiling by array</td>\n",
       "      <td>BioProject: https://www.ncbi.nlm.nih.gov/biopr...</td>\n",
       "      <td>1314</td>\n",
       "      <td>None</td>\n",
       "      <td>NIAID Group A Streptococcus</td>\n",
       "      <td>GPL1338</td>\n",
       "      <td>spotted DNA/cDNA</td>\n",
       "      <td>non-commercial</td>\n",
       "      <td>...</td>\n",
       "      <td>RNA</td>\n",
       "      <td>GSM26681</td>\n",
       "      <td>2</td>\n",
       "      <td>wt JRS4</td>\n",
       "      <td>160490</td>\n",
       "      <td>Streptococcus pyogenes M1 GAS</td>\n",
       "      <td>NaN</td>\n",
       "      <td>160490.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>None</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>8895 rows × 25 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                              GSE_title GSE_geo_accession  \\\n",
       "0     Temporal transcriptome analysis of Bacillus su...         GSE190460   \n",
       "1     Temporal transcriptome analysis of Bacillus su...         GSE190460   \n",
       "2     Temporal transcriptome analysis of Bacillus su...         GSE190460   \n",
       "3     Temporal transcriptome analysis of Bacillus su...         GSE190460   \n",
       "4     Temporal transcriptome analysis of Bacillus su...         GSE190460   \n",
       "...                                                 ...               ...   \n",
       "8890        M1 and M6 serotypes, Wild-type vs. Mga- GAS           GSE1895   \n",
       "8891  Wild-type and irr mutant Group A Streptococcus...           GSE1550   \n",
       "8892  Wild-type and irr mutant Group A Streptococcus...           GSE1550   \n",
       "8893  Wild-type and irr mutant Group A Streptococcus...           GSE1550   \n",
       "8894  Wild-type and irr mutant Group A Streptococcus...           GSE1550   \n",
       "\n",
       "                           GSE_type  \\\n",
       "0     Expression profiling by array   \n",
       "1     Expression profiling by array   \n",
       "2     Expression profiling by array   \n",
       "3     Expression profiling by array   \n",
       "4     Expression profiling by array   \n",
       "...                             ...   \n",
       "8890  Expression profiling by array   \n",
       "8891  Expression profiling by array   \n",
       "8892  Expression profiling by array   \n",
       "8893  Expression profiling by array   \n",
       "8894  Expression profiling by array   \n",
       "\n",
       "                                           GSE_relation   TaxId  \\\n",
       "0     BioProject: https://www.ncbi.nlm.nih.gov/biopr...  224308   \n",
       "1     BioProject: https://www.ncbi.nlm.nih.gov/biopr...  224308   \n",
       "2     BioProject: https://www.ncbi.nlm.nih.gov/biopr...  224308   \n",
       "3     BioProject: https://www.ncbi.nlm.nih.gov/biopr...  224308   \n",
       "4     BioProject: https://www.ncbi.nlm.nih.gov/biopr...  224308   \n",
       "...                                                 ...     ...   \n",
       "8890  BioProject: https://www.ncbi.nlm.nih.gov/biopr...    1314   \n",
       "8891  BioProject: https://www.ncbi.nlm.nih.gov/biopr...    1314   \n",
       "8892  BioProject: https://www.ncbi.nlm.nih.gov/biopr...    1314   \n",
       "8893  BioProject: https://www.ncbi.nlm.nih.gov/biopr...    1314   \n",
       "8894  BioProject: https://www.ncbi.nlm.nih.gov/biopr...    1314   \n",
       "\n",
       "                     GSE_supplementary_file  \\\n",
       "0     geo/series/GSE190nnn/GSE190460/suppl/   \n",
       "1     geo/series/GSE190nnn/GSE190460/suppl/   \n",
       "2     geo/series/GSE190nnn/GSE190460/suppl/   \n",
       "3     geo/series/GSE190nnn/GSE190460/suppl/   \n",
       "4     geo/series/GSE190nnn/GSE190460/suppl/   \n",
       "...                                     ...   \n",
       "8890      geo/series/GSE1nnn/GSE1895/suppl/   \n",
       "8891                                   None   \n",
       "8892                                   None   \n",
       "8893                                   None   \n",
       "8894                                   None   \n",
       "\n",
       "                                     GPL_title GPL_geo_accession  \\\n",
       "0     BaSysBio Bacillus subtilis T3 400K array          GPL21981   \n",
       "1     BaSysBio Bacillus subtilis T3 400K array          GPL21981   \n",
       "2     BaSysBio Bacillus subtilis T3 400K array          GPL21981   \n",
       "3     BaSysBio Bacillus subtilis T3 400K array          GPL21981   \n",
       "4     BaSysBio Bacillus subtilis T3 400K array          GPL21981   \n",
       "...                                        ...               ...   \n",
       "8890                     M1, M3, M18 GAS Array           GPL1482   \n",
       "8891               NIAID Group A Streptococcus           GPL1338   \n",
       "8892               NIAID Group A Streptococcus           GPL1338   \n",
       "8893               NIAID Group A Streptococcus           GPL1338   \n",
       "8894               NIAID Group A Streptococcus           GPL1338   \n",
       "\n",
       "               GPL_technology   GPL_distribution  ... GSM_type  \\\n",
       "0     in situ oligonucleotide  custom-commercial  ...      RNA   \n",
       "1     in situ oligonucleotide  custom-commercial  ...      RNA   \n",
       "2     in situ oligonucleotide  custom-commercial  ...      RNA   \n",
       "3     in situ oligonucleotide  custom-commercial  ...      RNA   \n",
       "4     in situ oligonucleotide  custom-commercial  ...      RNA   \n",
       "...                       ...                ...  ...      ...   \n",
       "8890  spotted oligonucleotide     non-commercial  ...      RNA   \n",
       "8891         spotted DNA/cDNA     non-commercial  ...      RNA   \n",
       "8892         spotted DNA/cDNA     non-commercial  ...      RNA   \n",
       "8893         spotted DNA/cDNA     non-commercial  ...      RNA   \n",
       "8894         spotted DNA/cDNA     non-commercial  ...      RNA   \n",
       "\n",
       "     GSM_geo_accession GSM_channel_count  \\\n",
       "0           GSM5724096                 1   \n",
       "1           GSM5724097                 1   \n",
       "2           GSM5724098                 1   \n",
       "3           GSM5724099                 1   \n",
       "4           GSM5724100                 1   \n",
       "...                ...               ...   \n",
       "8890          GSM33613                 2   \n",
       "8891          GSM26678                 2   \n",
       "8892          GSM26679                 2   \n",
       "8893          GSM26680                 2   \n",
       "8894          GSM26681                 2   \n",
       "\n",
       "                              GSM_source_name_ch1 GSM_taxid_ch1  \\\n",
       "0                         Bacillus subtilis NDmed          1423   \n",
       "1                         Bacillus subtilis NDmed          1423   \n",
       "2                         Bacillus subtilis NDmed          1423   \n",
       "3                         Bacillus subtilis NDmed          1423   \n",
       "4                         Bacillus subtilis NDmed          1423   \n",
       "...                                           ...           ...   \n",
       "8890  CsCl total RNA from M1 serotype GAS KSM165L          1314   \n",
       "8891                                      wt JRS4        160490   \n",
       "8892                                      wt JRS4        160490   \n",
       "8893                                      wt JRS4        160490   \n",
       "8894                                      wt JRS4        160490   \n",
       "\n",
       "                   GSM_organism_ch1  \\\n",
       "0                 Bacillus subtilis   \n",
       "1                 Bacillus subtilis   \n",
       "2                 Bacillus subtilis   \n",
       "3                 Bacillus subtilis   \n",
       "4                 Bacillus subtilis   \n",
       "...                             ...   \n",
       "8890         Streptococcus pyogenes   \n",
       "8891  Streptococcus pyogenes M1 GAS   \n",
       "8892  Streptococcus pyogenes M1 GAS   \n",
       "8893  Streptococcus pyogenes M1 GAS   \n",
       "8894  Streptococcus pyogenes M1 GAS   \n",
       "\n",
       "                                GSM_characteristics_ch1  GSM_taxid_ch2  \\\n",
       "0     strain: NDmed;genotype: Wild-type, undomestica...            NaN   \n",
       "1     strain: NDmed;genotype: Wild-type, undomestica...            NaN   \n",
       "2     strain: NDmed;genotype: Wild-type, undomestica...            NaN   \n",
       "3     strain: NDmed;genotype: Wild-type, undomestica...            NaN   \n",
       "4     strain: NDmed;genotype: Wild-type, undomestica...            NaN   \n",
       "...                                                 ...            ...   \n",
       "8890                                                NaN         1314.0   \n",
       "8891                                                NaN       160490.0   \n",
       "8892                                                NaN       160490.0   \n",
       "8893                                                NaN       160490.0   \n",
       "8894                                                NaN       160490.0   \n",
       "\n",
       "     GSM_characteristics_ch2                    GSM_supplementary_file  \n",
       "0                        NaN  geo/samples/GSM5724nnn/GSM5724096/suppl/  \n",
       "1                        NaN  geo/samples/GSM5724nnn/GSM5724097/suppl/  \n",
       "2                        NaN  geo/samples/GSM5724nnn/GSM5724098/suppl/  \n",
       "3                        NaN  geo/samples/GSM5724nnn/GSM5724099/suppl/  \n",
       "4                        NaN  geo/samples/GSM5724nnn/GSM5724100/suppl/  \n",
       "...                      ...                                       ...  \n",
       "8890                     NaN      geo/samples/GSM33nnn/GSM33613/suppl/  \n",
       "8891                     NaN                                      None  \n",
       "8892                     NaN                                      None  \n",
       "8893                     NaN                                      None  \n",
       "8894                     NaN                                      None  \n",
       "\n",
       "[8895 rows x 25 columns]"
      ]
     },
     "execution_count": 151,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Filtrado a nivel de especie \n",
    "filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 152,
   "metadata": {},
   "outputs": [],
   "source": [
    "species = set(filtered.GSM_geo_accession)\n",
    "identical = set(filtrado.GSM_geo_accession)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 153,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "8224"
      ]
     },
     "execution_count": 153,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(species)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 154,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6985"
      ]
     },
     "execution_count": 154,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(identical)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 155,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1.0"
      ]
     },
     "execution_count": 155,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(identical & species) / len(identical)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "base",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
